cpp/ql/src/Likely Bugs/Memory Management/PointerOverflow.ql

```diff
@@ -0,0 +1,31 @@
+/**
+ * @name Pointer overflow check
+ * @description Adding a value to a pointer to see if it overflows relies
+ *              on undefined behavior and may lead to memory corruption.
+ * @kind problem
+ * @problem.severity error
+ * @precision high
+ * @id cpp/pointer-overflow-check
+ * @tags reliability
+ *       security
+ */
+
+import cpp
+private import semmle.code.cpp.valuenumbering.GlobalValueNumbering
+private import semmle.code.cpp.commons.Exclusions
+
+from RelationalOperation ro, PointerAddExpr add, Expr expr1, Expr expr2
+where
+  ro.getAnOperand() = add and
+  add.getAnOperand() = expr1 and
+  ro.getAnOperand() = expr2 and
+  globalValueNumber(expr1) = globalValueNumber(expr2) and
+  // Exclude macros but not their arguments
+  not isFromMacroDefinition(ro) and
```

Comment:
`SignedOverflowCheck.ql` also checks that `exprMightOverflowPositively(add)` here (i.e. a range analysis check).  Is there a difference that justifies that check existing in one but not both versions of the query?

Reply:
The range analysis doesn't work for pointers, only numeric types. But the real reason is that the check suppresses some FPs that appeared because of implicit casts between integer types of varying width and signedness. Pointer types are all the same size in practice (let's ignore `near` and `far` pointers), so these kinds of FP don't appear.